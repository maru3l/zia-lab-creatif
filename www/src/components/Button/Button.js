// vendors
import React from "react"
import { Link } from "gatsby"
import { css } from "@emotion/core"
import { fluidRange } from "polished"

// styles
import { colors, transition } from "../../styles/variables"
import { useLayoutEffect } from "react"
import { useState } from "react"

const components = {
  link: Link,
  a: React.forwardRef(({ children, ...rest }, ref) => (
    <a ref={ref} {...rest}>
      {children}
    </a>
  )),
  href: React.forwardRef(({ children, ...rest }, ref) => (
    <a ref={ref} {...rest}>
      {children}
    </a>
  )),
  button: React.forwardRef(({ children, ...rest }, ref) => (
    <button ref={ref} {...rest}>
      {children}
    </button>
  )),
}

const getRGB = str => {
  return str
    .substring(str.lastIndexOf("(") + 1, str.lastIndexOf(")"))
    .split(", ")
    .map(x => Number(x))
}

const rgbToHex = (r, g, b) =>
  "#" +
  [Number(r), Number(g), Number(b)]
    .map(x => {
      const hex = x.toString(16)
      return hex.length === 1 ? "0" + hex : hex
    })
    .join("")

const getAncesterCssValue = (element, params = []) => {
  const styles = getComputedStyle(element)

  const results = params.map(key => {
    if (key === "background-color" && styles[key] === "rgba(0, 0, 0, 0)")
      return undefined

    return styles[key]
  })

  if (results.includes(undefined)) {
    const newParams = []

    results.forEach((value, index) => {
      if (typeof value === "undefined") {
        newParams.push(params[index])
      }
    })

    getAncesterCssValue(element.parentNode, newParams).forEach((value, i) => {
      const index = params.indexOf(newParams[i])

      results[index] = value
    })
  }

  return results
}

const Button = ({ children, tag, to, fitWithBgColor, ...rest }) => {
  const buttonRef = React.createRef()
  const [color, setColor] = useState(colors.doublePearlLusta)
  const [underlineColor, setUnderlineColor] = useState(colors.scarlet)
  const Tag = components[tag || "button"]

  const props = {
    to: tag === "link" ? to : undefined,
    href: tag === `href` || tag === `a` ? to : undefined,
    ...rest,
  }

  useLayoutEffect(() => {
    if (fitWithBgColor) {
      const response = getAncesterCssValue(buttonRef.current.parentNode, [
        "background-color",
        "color",
      ])

      const [bgColor, fontColor] = response

      const fontColorRGB = getRGB(fontColor)
      const fontColorHex = rgbToHex(
        fontColorRGB[0],
        fontColorRGB[1],
        fontColorRGB[2]
      )

      const bgColorRGB = getRGB(bgColor)
      const bgColorHex = rgbToHex(bgColorRGB[0], bgColorRGB[1], bgColorRGB[2])

      setColor(fontColorHex)

      if (
        fontColorHex === colors.scarlet &&
        bgColorHex === colors.doublePearlLusta
      ) {
        setUnderlineColor(colors.prussianBlue)
      }

      if (
        bgColorHex === colors.scarlet &&
        fontColorHex === colors.doublePearlLusta
      ) {
        setUnderlineColor(colors.prussianBlue)
      }

      if (
        bgColorHex === colors.prussianBlue &&
        fontColorHex === colors.doublePearlLusta
      ) {
        setUnderlineColor(colors.scarlet)
      }
    }
  }, [buttonRef, fitWithBgColor])

  const buttonStyles = css`
    font-family: inherit;
    cursor: pointer;
    display: inline-block;
    text-decoration: none;
    background: none;
    appearance: none;
    border: 0;
    font-size: ${41 / 10}rem;
    letter-spacing: ${10 / 1000}em;
    color: ${color};
    position: relative;
    z-index: 1;
    padding: 0 0.02em;
    margin: ${6.715 / 41}em 0;

    ${fluidRange(
      {
        prop: "font-size",
        fromSize: `${34 / 10}rem`,
        toSize: `${41 / 10}rem`,
      },
      "320px",
      "768px"
    )}

    &:before {
      background-color: ${underlineColor};
      /* height: calc(${6.715 / 41}em + 100%); */
      height: 100%;
      width: 100%;
      display: block;
      content: "";
      position: absolute;
      top: 0;
      left: 0;
      z-index: -1;
      transform: scaleY(0);
      transform-origin: bottom;
      transition: transform ${transition.speed.default}
        ${transition.curve.default};
    }

    &:after {
      background-color: ${underlineColor};
      /* height: ${6.715 / 41}em; */
      height: 9px;
      width: 100%;
      display: block;
      content: "";
      position: absolute;
      top: 99%;
      left: 0;
    }

    :hover,
    :focus {
      &:before {
        transform: scaleY(1);
        outline: 0;
      }
    }
  `

  const style = {
    "&&": {
      ...buttonStyles,
    },
  }

  return (
    <Tag css={style} {...props} ref={buttonRef}>
      {children}
    </Tag>
  )
}

export default Button
