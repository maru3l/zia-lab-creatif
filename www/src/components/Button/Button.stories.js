// vendors
import React from "react"
import { storiesOf } from "@storybook/react"
import { withKnobs, color, boolean } from "@storybook/addon-knobs"

import Button from "./index"

const stories = storiesOf("Button", module)

stories.addDecorator(withKnobs)

stories.add("Primary", () => (
  <Button
    disabled={boolean("Disabled", false)}
    parentBackgroundColor={color("Color of the parent background", "#ffffff")}
  >
    Voir les régions
  </Button>
))

stories.add("Secondary", () => (
  <Button
    secondary
    disabled={boolean("Disabled", false)}
    parentBackgroundColor={color("Color of the parent background", "#ffffff")}
  >
    Voir les régions
  </Button>
))

stories.add("Primary - 2 line", () => (
  <Button
    disabled={boolean("Disabled", false)}
    parentBackgroundColor={color("Color of the parent background", "#ffffff")}
  >
    Remplir le formulaire de demande de stage
  </Button>
))

stories.add("Secondary - 2 line", () => (
  <Button
    secondary
    disabled={boolean("Disabled", false)}
    parentBackgroundColor={color("Color of the parent background", "#ffffff")}
  >
    Remplir le formulaire de demande de stage
  </Button>
))

stories.add("Primary Full", () => (
  <Button
    full
    disabled={boolean("Disabled", false)}
    parentBackgroundColor={color("Color of the parent background", "#ffffff")}
  >
    Voir les régions
  </Button>
))

stories.add("Secondary Full", () => (
  <Button
    full
    secondary
    disabled={boolean("Disabled", false)}
    parentBackgroundColor={color("Color of the parent background", "#ffffff")}
  >
    Voir les régions
  </Button>
))

stories.add("Primary Full - 2 line", () => (
  <Button
    full
    disabled={boolean("Disabled", false)}
    parentBackgroundColor={color("Color of the parent background", "#ffffff")}
  >
    Remplir le formulaire de demande de stage
  </Button>
))

stories.add("Secondary Full - 2 line", () => (
  <Button
    full
    secondary
    disabled={boolean("Disabled", false)}
    parentBackgroundColor={color("Color of the parent background", "#ffffff")}
  >
    Remplir le formulaire de demande de stage
  </Button>
))
