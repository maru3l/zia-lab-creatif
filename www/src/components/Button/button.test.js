import React from "react"
import renderer from "react-test-renderer"
import { render, cleanup, fireEvent } from "@testing-library/react"

import Button from "./Button"

afterEach(cleanup)

describe("Selective tag rendering", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Button>Voir les r√©gions</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("Defaults to rendering a button", () => {
    const { container, getByText } = render(
      <Button onClick={() => {}}>Hello</Button>
    )

    expect(container.nodeName).toBe(`DIV`)
    expect(getByText("Hello").nodeName).toBe("BUTTON")
  })

  it("renders a button if tag=button", () => {
    const { getByText } = render(
      <Button tag="button" onClick={() => {}}>
        Hello
      </Button>
    )

    expect(getByText("Hello").nodeName).toBe("BUTTON")
  })

  it("renders a button if tag=a", () => {
    const { getByText } = render(
      <Button tag="a" to="/somewhere">
        Hello
      </Button>
    )

    expect(getByText("Hello").nodeName).toBe("A")
  })

  it("renders a button if tag=link", () => {
    const { getByText } = render(
      <Button tag="link" to="/somewhere">
        Hello
      </Button>
    )

    expect(getByText("Hello").nodeName).toBe("A")
    expect(getByText("Hello").getAttribute("href")).toBe("/somewhere")
  })

  it("renders a link if tag=href", () => {
    const { getByText } = render(
      <Button tag="href" to="/somewhere">
        Hello
      </Button>
    )

    expect(getByText("Hello").nodeName).toBe("A")
    expect(getByText("Hello").getAttribute("href")).toBe("/somewhere")
  })
})

describe(`props forwarding`, () => {
  it(`invokes custom onClick handler`, () => {
    const onClick = jest.fn()

    const { getByText } = render(
      <Button to="/custom-click" onClick={onClick}>
        Hello
      </Button>
    )
    fireEvent.click(getByText(`Hello`))

    expect(onClick).toHaveBeenCalledWith(expect.any(Object))
  })

  it(`renders children`, () => {
    const children = <span>Hello World</span>
    const { getByText } = render(<Button to="/children">{children}</Button>)

    expect(getByText(`Hello World`)).toBeDefined()
  })
})
