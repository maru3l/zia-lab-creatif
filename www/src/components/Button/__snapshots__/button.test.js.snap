// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selective tag rendering renders correctly 1`] = `
<button
  css={
    Object {
      "&&": Object {
        ":focus": Object {
          "boxShadow": "0 0 0 10 #00353d",
          "outline": 0,
        },
        ":hover, :focus": Object {
          ".eqrrkvw0": Object {
            "transform": "translateX(10px)",
          },
          "borderColor": "#00353d",
          "color": "#00353d",
        },
        "MozOsxFontSmoothing": "grayscale",
        "WebkitFontSmoothing": "antialiased",
        "alignItems": "right",
        "backgroundColor": "transparent",
        "borderColor": "#e86752",
        "borderRadius": "13% / 50%",
        "borderStyle": "solid",
        "borderWidth": 3,
        "color": "#e86752",
        "cursor": "pointer",
        "display": "inline-block",
        "flexShrink": 0,
        "fontFamily": "Source Serif Pro",
        "fontSize": 24,
        "fontWeight": 700,
        "lineHeight": 1.08,
        "maxWidth": "26ch",
        "padding": "20px 40px 17px",
        "textAlign": "left",
        "textDecoration": "none",
        "textRendering": "optimizeLegibility",
        "transition": "all 250ms",
      },
    }
  }
>
  Voir les régions
  <span
    className="css-18hem83-Arrow eqrrkvw0"
  >
    →
  </span>
</button>
`;
